server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -171,6 +172,44 @@ public DataStream promoteDataStream() {
         return new DataStream(name, timeStampField, indices, getGeneration(), metadata, hidden, false);
     }
 
+    /**
+     * Reconciles this data stream with its state at the end of the snapshot.
+     *
+     * @param afterSnapshot The post-snapshot data stream
+     * @return Reconciled {@link DataStream} instance
+     */
+    public DataStream reconcile(DataStream afterSnapshot) {
+        if (name.equals(afterSnapshot.name) == false ||
+            timeStampField.equals(afterSnapshot.timeStampField) == false ||
+            hidden != afterSnapshot.hidden ||
+            replicated != afterSnapshot.replicated) {
+            throw new IllegalArgumentException(
+                "cannot reconcile data streams with differing names, timestamp fields, hidden statuses, or replicated statuses"
+            );
+        }
+
+        // starting from the pre-snapshot state ensures that any indices added during snapshot are excluded
+        List<Index> reconciledIndices = new ArrayList<>(this.indices);
+        // do not include indices deleted during snapshot
+        reconciledIndices.removeIf(x -> afterSnapshot.indices.contains(x) == false);
+
+        if (reconciledIndices.size() == 0) {
+            throw new IllegalArgumentException("cannot reconcile data streams in which all pre-snapshot backing indices were deleted");
+        }
```

Comment:
I wasn't sure if we should also reject the deletion of the pre-snapshot data stream's write index. If we do not reject that, the data stream restored in the snapshot will end up with a different write index.

Reply:
I think we shouldn't reject, but just use the indices' provided by `afterSnapshot`?